tests/trainer/connectors/test_data_connector.py

```diff
@@ -71,6 +72,84 @@ def test_dataloader_source_request_from_module():
+)
+def test_datahook_selector(hook_name):
+    def overridden_func(batch, *args, **kwargs):
+        return batch
+
+    def _reset_instances():
+        return BoringDataModule(), BoringModel(), Trainer()
+
+    def _no_datamodule_no_overridden():
+        model, _, trainer = _reset_instances()
+        trainer._data_connector.attach_datamodule(model, datamodule=None)
+        with no_warning_call(match="have overridden `{hook_name}` in both"):
+            hook = trainer._data_connector._datahook_selector.get_hook(hook_name)
+
+        assert hook == getattr(model, hook_name)
+
+    def _with_datamodule_no_overridden():
+        model, dm, trainer = _reset_instances()
+        trainer._data_connector.attach_datamodule(model, datamodule=dm)
+        with no_warning_call(match="have overridden `{hook_name}` in both"):
+            hook = trainer._data_connector._datahook_selector.get_hook(hook_name)
+
+        assert hook == getattr(model, hook_name)
+
+    def _override_model_hook():
+        model, dm, trainer = _reset_instances()
+        trainer._data_connector.attach_datamodule(model, datamodule=dm)
+        with no_warning_call(match="have overridden `{hook_name}` in both"):
+            hook = trainer._data_connector._datahook_selector.get_hook(hook_name)
+
+        assert hook == getattr(model, hook_name)
+
+    def _override_datamodule_hook():
+        model, dm, trainer = _reset_instances()
+        trainer._data_connector.attach_datamodule(model, datamodule=dm)
+        setattr(dm, hook_name, overridden_func)
+        with no_warning_call(match="have overridden `{hook_name}` in both"):
+            hook = trainer._data_connector._datahook_selector.get_hook(hook_name)
+
+        assert hook == getattr(dm, hook_name)
+
+    def _override_both_model_and_datamodule():
+        model, dm, trainer = _reset_instances()
+        trainer._data_connector.attach_datamodule(model, datamodule=dm)
+        setattr(model, hook_name, overridden_func)
+        setattr(dm, hook_name, overridden_func)
+        with pytest.warns(UserWarning, match=f"have overridden `{hook_name}` in both"):
+            hook = trainer._data_connector._datahook_selector.get_hook(hook_name)
+
+        warning_cache.clear()
+        assert hook == getattr(dm, hook_name)
+
+    def _with_datamodule_override_model():
+        model, dm, trainer = _reset_instances()
+        trainer._data_connector.attach_datamodule(model, datamodule=dm)
+        setattr(model, hook_name, overridden_func)
+        with pytest.warns(UserWarning, match=f"have overridden `{hook_name}` in `LightningModule`"):
+            hook = trainer._data_connector._datahook_selector.get_hook(hook_name)
+
+        warning_cache.clear()
+        assert hook == getattr(model, hook_name)
+
+    _no_datamodule_no_overridden()
+    _with_datamodule_no_overridden()
+    _override_model_hook()
+    _override_datamodule_hook()
+    _override_both_model_and_datamodule()
+    _with_datamodule_override_model()
```

Comment:
Why not write 6 separate tests instead of 1 that calls 6 local functions?

Or a testing class: https://docs.pytest.org/en/6.2.x/getting-started.html#group-multiple-tests-in-a-class